pkg/install/install.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2019 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package install
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/heptio/velero/pkg/client"
+)
+
+// kindToResource translates a Kind (mixed case, singular) to a Resource (lowercase, plural) string.
+// This is to accomodate the dynamic client's need for an APIResource, as the Unstructured objects do not have easy helpers for this information.
+var kindToResource = map[string]string{
+	"CustomResourceDefinition": "customresourcedefinitions",
+	"Namespace":                "namespaces",
+	"ClusterRoleBinding":       "clusterrolebindings",
+	"ServiceAccount":           "serviceaccounts",
+	"Deployment":               "deployments",
+	"DaemonSet":                "daemonsets",
+	"Secret":                   "secrets",
+	"BackupStorageLocation":    "backupstoragelocations",
+	"VolumeSnapshotLocation":   "volumesnapshotlocations",
+}
+
+// Install creates resources on the Kubernetes cluster.
+// An unstructured list of resources is sent, one at a time, to the server. These are assumed to be in the preferred order already.
+// An io.Writer can be used to output to a log or the console.
+func Install(factory client.DynamicFactory, resources *unstructured.UnstructuredList, w io.Writer) error {
```

Comment:
I'm thinking that when you add waiting for CRDs to be ready, you'll want this function to accept a slice of structs like:

```go
type ResourceGroup struct {
    Resources []*unstructured.Unstructured
    Ready func(dynamicFactory client.DynamicFactory) (bool, error)
}
```

So you can have one `ResourceGroup` for all the CRDs, and the Ready func uses the dynamic factory to check them for readiness. The `Install` func will loop through each `ResourceGroup`, and for each, create all of the `Resources`, then call the `Ready` func in a polling loop, up to a certain period. Not dissimilar from the "wait for terminated" stuff in restore.go.  This is my initial thought off the top of my head but obv can be refined/changed.

Reply:
Hmm...so in this case, we'd need a group for each type? CRDs have the most, while the rest are 1 object. This also means creating different structs for printing and submitting to the API, right? Our current printer doesn't like using `[]*unstructured.Unstructured` directly, instead needing a list.

I do prefer the idea of doing a wait vs sleep, though.

Reply:
You don't need a group per type - you need the CRDs in one group with an appropriate `Ready`, and then any CR instances in a second group.  All the rest of the stuff could go into the CRD group since they're not dependent on the CRDs being ready.

You can make Resources an unstructuredlist, that's not an issue